packages/blob-index/src/sharded-dag-index.js

```diff
@@ -0,0 +1,251 @@
+  if (!rootBytes) {
+    return error(new DecodeFailure(`missing root block: ${root}`))
+  }
+
+  const [version, dagIndexData] = ShardedDAGIndexSchema.match(
+    dagCBOR.decode(rootBytes)
+  )
+  switch (version) {
+    case 'index/sharded/dag@0.1': {
+      const dagIndex = {
+        content: dagIndexData.content,
+        shards: new DigestMap(),
+      }
+      for (const shard of dagIndexData.shards) {
+        const shardBytes = blocks.get(shard.multihash)
+        if (!shardBytes) {
+          return error(new DecodeFailure(`missing shard block: ${shard}`))
+        }
+
+        const blobIndexData = BlobIndexSchema.from(dagCBOR.decode(shardBytes))
+        const blobIndex = new DigestMap()
+        for (const [digest, [offset, length]] of blobIndexData[1]) {
+          blobIndex.set(Digest.decode(digest), [offset, length])
+        }
+        dagIndex.shards.set(Digest.decode(blobIndexData[0]), blobIndex)
+      }
+      return ok(dagIndex)
+    }
+    default:
+      return error(
+        /** @type {import('@web3-storage/capabilities/types').UnknownFormat} */
+        ({ name: 'UnknownFormat' })
+      )
+  }
+}
+
+class DecodeFailure extends Failure {
+  #reason
+
+  /** @param {string} [reason] */
+  constructor(reason) {
+    super()
+    this.name = /** @type {const} */ ('DecodeFailure')
+    this.#reason = reason
+  }
+
+  describe() {
+    return this.#reason ?? 'failed to decode'
+  }
+}
+
+/**
+ *
+ * @param {API.UnknownLink} content
+ * @returns {API.ShardedDAGIndex}
+ */
+export function createShardedDAGIndex(content) {
+  return new ShardedDAGIndex(content)
+}
+
+/** @implements {API.ShardedDAGIndex} */
+class ShardedDAGIndex {
+  /** @param {API.UnknownLink} content */
+  constructor(content) {
+    this.content = content
+    this.shards = /** @type {API.ShardedDAGIndex['shards']} */ (new DigestMap())
+  }
+
+  /** @returns {Promise<API.Result<Uint8Array>>} */
+  async toArchive() {
+    const blocks = new Map()
+    const shards = [...this.shards.entries()].sort((a, b) =>
+      compare(a[0].digest, b[0].digest)
+    )
+    const index = {
+      content: this.content,
+      shards: /** @type {UI.Link[]} */ ([]),
+    }
+    for (const s of shards) {
+      const slices = [...s[1].entries()]
+        .sort((a, b) => compare(a[0].digest, b[0].digest))
+        .map((e) => [e[0].bytes, e[1]])
+      const bytes = dagCBOR.encode([s[0].bytes, slices])
+      const digest = await sha256.digest(bytes)
+      const cid = Link.create(dagCBOR.code, digest)
+      blocks.set(cid.toString(), { cid, bytes })
+      index.shards.push(cid)
+    }
+    const bytes = dagCBOR.encode({ 'index/sharded/dag@0.1': index })
+    const digest = await sha256.digest(bytes)
+    const cid = Link.create(dagCBOR.code, digest)
+    return ok(CAR.encode({ roots: [{ cid, bytes }], blocks }))
+  }
+}
+
+/**
+ * @param {Uint8Array} carData
+ * @returns {Promise<ShardedDAGIndex>}
+ */
+export const fromArchive = async (carData) => {
```

Comment:
This is what `extract` does above no?

Reply:
Extract expects to read from a `ReadableStream` whereas `fromArchive` reads from a `UInt8Array` in memory. Does this mean that `fromArchive` should just convert the `Uint8Array` into a `ReadableStream`?

Also, `extract` will not know how to do things like discarding the overall dag shard from the encoded shards, before building the ShardedDAGIndex.